src/transform-sdk/rust/sr/src/lib.rs

```diff
@@ -80,10 +87,176 @@ impl Default for SchemaRegistryClient {
     }
 }
 
+#[deprecated(
+    since = "1.0.3",
+    note = "Prefer decode_schema_id to receive both the extracted ID and a byte slice to the rest of the buffer"
+)]
 pub fn extract_id(buf: &[u8]) -> Result<SchemaId> {
     static MAGIC_BYTES: [u8; 1] = [0x00];
     if !buf.starts_with(&MAGIC_BYTES) || buf.len() < 5 {
         return Err(SchemaRegistryError::BadHeader);
     }
     Ok(SchemaId(i32::from_be_bytes(buf[1..5].try_into().unwrap())))
 }
+
+pub fn decode_schema_id(buf: &[u8]) -> Result<(SchemaId, &[u8])> {
+    static MAGIC_BYTES: [u8; 1] = [0x00];
+    if !buf.starts_with(&MAGIC_BYTES) || buf.len() < 5 {
+        return Err(SchemaRegistryError::BadHeader);
+    }
+    Ok((
+        SchemaId(i32::from_be_bytes(buf[1..5].try_into().unwrap())),
+        &buf[5..],
+    ))
+}
+
+pub fn encode_schema_id(id: SchemaId, buf: &[u8]) -> Vec<u8> {
+    static MAGIC_BYTES: [u8; 1] = [0x00];
+    let id_bytes = id.0.to_be_bytes();
+    [&MAGIC_BYTES, &id_bytes[0..], buf].concat().to_vec()
+}
+
+#[derive(Debug)]
+struct LruCache<K: HashT + Eq, V: Clone> {
+    underlying: RefCell<LruCacheImpl<K, V>>,
+}
+
+#[allow(dead_code)]
+impl<Key: HashT + Eq, Value: Clone> LruCache<Key, Value> {
+    pub fn new(max_entries: Option<usize>) -> LruCache<Key, Value> {
+        let nz_max = match max_entries {
+            Some(v) if v > 0 => NonZeroUsize::new(v),
+            _ => NonZeroUsize::new(10),
+        };
+        Self {
+            underlying: RefCell::new(LruCacheImpl::<Key, Value>::new(nz_max.unwrap())),
+        }
+    }
+
+    fn get(&self, k: &Key) -> Option<Value> {
+        if let Some(v) = self.underlying.borrow_mut().get(k) {
+            Some((*v).clone())
+        } else {
+            None
+        }
+    }
+
+    fn put(&self, k: Key, v: Value) -> Option<Value> {
+        self.underlying.borrow_mut().put(k, v)
+    }
+}
+
+#[derive(Debug)]
+struct CachingSchemaRegistryClient {
+    underlying: AbiSchemaRegistryClient,
+    schema_by_id_cache: LruCache<SchemaId, Schema>,
+    schema_by_sv_cache: LruCache<SubjectVersion, SubjectSchema>,
+}
+
+/// SubjectVersion is an aggregate type for the registered subject and version of a schema
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+struct SubjectVersion {
+    subject: String,
+    version: SchemaVersion,
+}
+
+impl SubjectVersion {
+    pub fn new(subject: impl Into<String>, version: SchemaVersion) -> Self {
```

Comment:
`pub` is not needed because `SubjectVersion` is not `pub` right?

Reply:
yup, meant to remove that

Reply:
done